server/src/infra/ldap_handler.rs

```diff
@@ -211,6 +224,216 @@ fn root_dse_response(base_dn: &str) -> LdapOp {
+            .join(" ")
+    }
+}
+
+// See RFC4512 section 4.2 "Subschema Subentries"
+// This struct holds all information on what attributes and objectclasses are present on the server.
+// It can be used to 'index' a server using a LDAP subschema call.
+pub struct LdapSchemaDescription {
+    base: Schema,
+    user_object_classes: ObjectClassList,
+    group_object_classes: ObjectClassList,
+}
+
+impl Deref for LdapSchemaDescription {
+    type Target = Schema;
+
+    fn deref(&self) -> &Self::Target {
+        &self.base
+    }
+}
+
+impl LdapSchemaDescription {
+    fn from(schema: &Schema) -> Self {
+        let mut user_object_classes = get_default_user_object_classes();
+        user_object_classes.extend(schema.extra_user_object_classes.clone());
+        let mut group_object_classes = get_default_group_object_classes();
+        group_object_classes.extend(schema.extra_group_object_classes.clone());
+
+        Self {
+            base: PublicSchema::from(schema.clone()).get_schema().clone(),
+            user_object_classes: ObjectClassList(user_object_classes),
+            group_object_classes: ObjectClassList(group_object_classes),
+        }
+    }
+
+    fn required_user_attributes(&self) -> AttributeList {
+        let attributes = self
+            .user_attributes
+            .attributes
+            .iter()
+            .filter(|&a| get_required_user_attributes().contains(&a.name))
+            .cloned()
+            .collect();
+
+        AttributeList { attributes }
+    }
+
+    fn optional_user_attributes(&self) -> AttributeList {
+        let attributes = self
+            .user_attributes
+            .attributes
+            .iter()
+            .filter(|&a| !get_required_user_attributes().contains(&a.name))
+            .cloned()
+            .collect();
+
+        AttributeList { attributes }
+    }
+
+    fn required_group_attributes(&self) -> AttributeList {
+        let attributes = self
+            .group_attributes
+            .attributes
+            .iter()
+            .filter(|&a| get_required_group_attributes().contains(&a.name))
+            .cloned()
+            .collect();
+
+        AttributeList { attributes }
+    }
+
+    fn optional_group_attributes(&self) -> AttributeList {
+        let attributes = self
+            .group_attributes
+            .attributes
+            .iter()
+            .filter(|&a| !get_required_group_attributes().contains(&a.name))
+            .cloned()
+            .collect();
+
+        AttributeList { attributes }
+    }
+
+    // See RFC4512 section 4.2.2 "attributeTypes"
+    fn formatted_attribute_list(&self) -> Vec<Vec<u8>> {
+        let mut formatted_list: Vec<Vec<u8>> = Vec::new();
+
+        for (index, attribute) in self.all_attributes().attributes.into_iter().enumerate() {
+            formatted_list.push(
+                format!(
+                    "( 2.{} NAME '{}' DESC 'LLDAP: {}' SUP {:?} )",
+                    (index + 4),
```

Comment:
What's that value 4? Can you extract it to a named constant so the reader knows where it's from?

Reply:
The constant 4 was an offset needed as the builtin attributes had already been added earlier. Extracted it to a parameter to the function, and passed it from the call-site (with a comment) where the builtin attributes are added, so the connection is clear.